### PR TITLE
Save walker weights in HDF5 files

### LINKED ISSUE #4157: Read/write walker weights in restart
**Is your feature request related to a problem? Please describe.**
Need to record walker weights.

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

Saves the walker weights following the same pattern used to save the particle positions.

This fixes #4157

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/Particle/HDFWalkerInput_0_4.cpp [hunk 1]
```diff
@@ -168,11 +169,17 @@ bool HDFWalkerInput_0_4::read_hdf5(const std::filesystem::path& h5name)
     const int nitems    = num_ptcls_ * OHMMS_DIM;
     const int curWalker = wc_list_.getActiveWalkers();
     wc_list_.createWalkers(nw_in, num_ptcls_);
-    Buffer_t::iterator it(posin.begin() + woffsets[myComm->rank()] * nitems);
+    Buffer_t::iterator it = posin.begin() + woffsets[myComm->rank()] * nitems;
+    auto it_w             = weights_in.begin() + woffsets[myComm->rank()];
```
Comment (ye-luo) [inline src/Particle/HDFWalkerInput_0_4.cpp:173]: Can we just compute it and it_w inside the loop? iterator is so hard to follow.

**Review by ye-luo — COMMENTED** [review]
Could you document the h5 data layout in the manual.

Mention the layout of offsets, electron positions and weights dataset.

**Comment by quantumsteve** [comment]
> Could you document the h5 data layout in the manual.

Is this the best place to document the `.config.h5` file? Is it documented elsewhere?
https://github.com/QMCPACK/qmcpack/blob/507d0de7dcdbee1b1714b57fb7b688e96586fb36/docs/output_overview.rst#the-configh5-file

File: src/Particle/HDFWalkerInput_0_4.cpp (hunk 1 above)
Comment (quantumsteve) [reply src/Particle/HDFWalkerInput_0_4.cpp]: I refactored those 3 loops so they're easier to read.

**Comment by ye-luo** [comment]
> > Could you document the h5 data layout in the manual.
> 
> Is this the best place to document the `.config.h5` file? Is it documented elsewhere? https://github.com/QMCPACK/qmcpack/blob/507d0de7dcdbee1b1714b57fb7b688e96586fb36/docs/output_overview.rst#the-configh5-file

I don't have a better place. Just do it there.

File: src/Particle/HDFWalkerInput_0_4.cpp [hunk 2]
```diff
@@ -147,10 +147,11 @@ bool HDFWalkerInput_0_4::read_hdf5(const std::filesystem::path& h5name)
   }
 
   using Buffer_t = std::vector<QMCTraits::RealType>;
-  Buffer_t posin;
   std::array<size_t, 3> dims{nw_in, num_ptcls_, OHMMS_DIM};
-  posin.resize(dims[0] * dims[1] * dims[2]);
+  Buffer_t posin(dims[0] * dims[1] * dims[2]);
   hin.readSlabReshaped(posin, dims, hdf::walkers);
+  std::vector<QMCTraits::FullPrecRealType> weights_in;
+  bool has_weights = hin.readEntry(weights_in, hdf::walker_weights);
```
Comment (ye-luo) [inline src/Particle/HDFWalkerInput_0_4.cpp:154]: Add const, if the value is intended to be read-only.

File: src/Particle/HDFWalkerInput_0_4.cpp [hunk 3]
```diff
@@ -168,11 +169,19 @@ bool HDFWalkerInput_0_4::read_hdf5(const std::filesystem::path& h5name)
...
     const int curWalker = wc_list_.getActiveWalkers();
     wc_list_.createWalkers(nw_in, num_ptcls_);
-    Buffer_t::iterator it(posin.begin() + woffsets[myComm->rank()] * nitems);
-    for (int i = 0, iw = curWalker; i < nw_in; ++i, ++iw)
+
+    auto it = posin.begin() + woffsets[myComm->rank()] * nitems;
+    for (int i = 0; i < nw_in; ++i, it += nitems)
+    {
+      copy(it, it + nitems, get_first_address(wc_list_[i + curWalker]->R));
+    }
```
Comment (ye-luo) [inline src/Particle/HDFWalkerInput_0_4.cpp:177]: Remove {} around 1 line.

File: src/Particle/HDFWalkerInput_0_4.cpp [hunk 4]
```diff
@@ -168,11 +169,19 @@ bool HDFWalkerInput_0_4::read_hdf5(const std::filesystem::path& h5name)
...
+    if (has_weights)
     {
-      copy(it, it + nitems, get_first_address(wc_list_[iw]->R));
-      it += nitems;
+      const auto woffset = woffsets[myComm->rank()];
+      for (int i = 0; i < nw_in; ++i)
+      {
+        wc_list_[i + curWalker]->Weight = weights_in[i + woffset];
```
Comment (ye-luo) [inline src/Particle/HDFWalkerInput_0_4.cpp:183]: Remove {} around 1 line.

File: src/Particle/HDFWalkerInput_0_4.cpp [hunk 5]
```diff
@@ -218,6 +227,11 @@ bool HDFWalkerInput_0_4::read_hdf5_scatter(const std::filesystem::path& h5name)
   std::vector<int> counts(myComm->size()), woffsets(np1, 0);
   FairDivideLow(nw_in, myComm->size(), woffsets);
 
+  auto woffsets_w = woffsets;
```
Comment (ye-luo) [inline src/Particle/HDFWalkerInput_0_4.cpp:230]: This code is very confusing. Prefer the following flow
1. define woffsets_w and FairDivideLow it.
2. setup counts_w
3. define woffsets and then loop assign each value.
4. setup counts.

About variable names. Seeing counts_w, I can understand that is for weights.
But for counts, we'd better to add a comment before its definition. "for electron coordinates"
Please do this for all the variables in this category.

File: src/Particle/HDFWalkerInput_0_4.cpp [hunk 6]
```diff
@@ -215,38 +220,53 @@ bool HDFWalkerInput_0_4::read_hdf5_scatter(const std::filesystem::path& h5name)
...
+  std::vector<int> woffsets_weights(np1, 0);
+  FairDivideLow(nw_in, myComm->size(), woffsets_weights);
 
-  const size_t nw_loc = woffsets[myComm->rank() + 1] - woffsets[myComm->rank()];
+  std::vector<int> counts_weights(myComm->size());
+  for (int i = 0; i < counts_weights.size(); ++i)
+    counts_weights[i] = woffsets_weights[i + 1] - woffsets_weights[i];
 
+  // walker counts and offsets for electron coordinates
+  std::vector<int> woffsets(woffsets_weights);
```
Comment (ye-luo) [inline src/Particle/HDFWalkerInput_0_4.cpp:231]: I meant here
```
std::vector<int> woffsets(np1);
```
and below do
```
woffsets[i] = woffsets_weights[i] * nitems;
```
I feel this is more readable.

Comment (quantumsteve) [reply src/Particle/HDFWalkerInput_0_4.cpp]: done

File: src/Particle/HDFWalkerInput_0_4.cpp (hunk 5 above)
Comment (quantumsteve) [reply src/Particle/HDFWalkerInput_0_4.cpp]: done

**Comment by quantumsteve** [comment]
> > > Could you document the h5 data layout in the manual.
> > 
> > 
> > Is this the best place to document the `.config.h5` file? Is it documented elsewhere? https://github.com/QMCPACK/qmcpack/blob/507d0de7dcdbee1b1714b57fb7b688e96586fb36/docs/output_overview.rst#the-configh5-file
> 
> I don't have a better place. Just do it there.

I dumped the contents from one of our restart tests. Still need to describe the individual groups and datasets.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]
